modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/checkpoint/CheckpointManager.java

```diff
@@ -0,0 +1,214 @@
+
+        checkpointWorkflow = new CheckpointWorkflow(
+                checkpointConfig,
+                checkpointMarkersStorage,
+                checkpointReadWriteLock,
+                dataRegions
+        );
+
+        checkpointPagesWriterFactory = new CheckpointPagesWriterFactory(
+                logger.apply(CheckpointPagesWriterFactory.class),
+                (fullPage, buf, tag) -> filePageStoreManager.write(fullPage.groupId(), fullPage.pageId(), buf, tag, true),
+                pageSize
+        );
+
+        checkpointer = new Checkpointer(
+                logger.apply(Checkpoint.class),
+                igniteInstanceName,
+                workerListener,
+                longJvmPauseDetector,
+                checkpointWorkflow,
+                checkpointPagesWriterFactory,
+                checkpointConfig
+        );
+
+        checkpointTimeoutLock = new CheckpointTimeoutLock(
+                logger.apply(CheckpointTimeoutLock.class),
+                checkpointReadWriteLock,
+                checkpointConfigView.readLockTimeout(),
+                () -> safeToUpdateAllPageMemories(dataRegions),
+                checkpointer
+        );
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void start() {
+        checkpointWorkflow.start();
+
+        checkpointer.start();
+
+        checkpointTimeoutLock.start();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void stop() throws Exception {
+        IgniteUtils.closeAll(
+                checkpointTimeoutLock::stop,
+                checkpointer::stop,
+                checkpointWorkflow::stop
+        );
+    }
+
+    /**
+     * Returns checkpoint timeout lock which can be used for protection of writing to memory.
+     */
+    public CheckpointTimeoutLock checkpointTimeoutLock() {
+        return checkpointTimeoutLock;
+    }
+
+    /**
+     * Adds a listener to be called for the corresponding persistent data region.
+     *
+     * @param listener Listener.
+     * @param dataRegion Persistent data region for which listener is corresponded to, {@code null} for all regions.
+     */
+    public void addCheckpointListener(CheckpointListener listener, PageMemoryDataRegion dataRegion) {
+        checkpointWorkflow.addCheckpointListener(listener, dataRegion);
+    }
+
+    /**
+     * Removes the listener.
+     *
+     * @param listener Listener.
+     */
+    public void removeCheckpointListener(CheckpointListener listener) {
+        checkpointWorkflow.removeCheckpointListener(listener);
+    }
+
+    /**
+     * Start the new checkpoint immediately.
+     *
+     * @param reason Checkpoint reason.
+     * @param finishListener Listener which will be called on finish checkpoint.
+     * @return Triggered checkpoint progress.
+     */
+    public CheckpointProgress forceCheckpoint(
+            String reason,
+            @Nullable BiConsumer<Void, Throwable> finishListener
```

Comment:
I still don't understand this. Don't we have checkpoint futures for various stages? Can't we just add listeners to them?